src/cache-fs.ts

```diff
@@ -124,53 +124,55 @@ export class CacheFileSystem implements FileSystem {
-        return this.fs.deleteFile(fullPath)
-            .then(() => this.cache.deleteFile(fullPath));
+    async deleteFile(fullPath:string): Promise<void> {
+        await this.fs.deleteFile(fullPath);
+        this.cache.deleteFileSync(fullPath);
     }
 
-    deleteDirectory(fullPath: string, recursive: boolean = false): Promise<void> {
-        return this.fs.deleteDirectory(fullPath, recursive)
-            .then(() => this.cache.deleteDirectory(fullPath, recursive));
+    async deleteDirectory(fullPath: string, recursive: boolean = false): Promise<void> {
+        await this.fs.deleteDirectory(fullPath, recursive);
+        this.cache.deleteDirectorySync(fullPath, recursive);
     }
 
-    ensureDirectory(fullPath:string): Promise<void> {
-        return this.fs.ensureDirectory(fullPath)
-            .then(() => this.cache.ensureDirectory(fullPath));
+    async ensureDirectory(fullPath:string): Promise<void> {
+        await this.fs.ensureDirectory(fullPath);
+        this.cache.ensureDirectorySync(fullPath);
     }
 
-    loadTextFile(fullPath: string): Promise<string> {
-        if (this.pathsInCache[fullPath]) return this.cache.loadTextFile(fullPath)
-        return this.fs.loadTextFile(fullPath)
-            .then(file => {
-                return this.cache.saveFile(fullPath, file)
-                    .then(() => this.pathsInCache[fullPath] = true)
-                    .then(() => this.loadTextFile(fullPath))
-            });
+    async loadTextFile(fullPath: string): Promise<string> {
+        if (this.pathsInCache[fullPath]) {
+            return this.cache.loadTextFileSync(fullPath);
+        }
+
+        const file = await this.fs.loadTextFile(fullPath);
+        this.cache.saveFileSync(fullPath, file);
+        return this.cache.loadTextFileSync(fullPath);
     }
 
-    loadDirectoryTree(fullPath?:string): Promise<Directory> {
-        if (this.isTreeCached) return this.cache.loadDirectoryTree(fullPath);
-        return this.cacheTree().then(() => {
-            this.isTreeCached = true;
-            return this.loadDirectoryTree(fullPath);
-        });
+    async loadDirectoryTree(fullPath?:string): Promise<Directory> {
+        if (this.isTreeCached) {
+            return this.cache.loadDirectoryTreeSync(fullPath);
+        }
+
+        await this.cacheTree();
+        this.isTreeCached = true;
+        return this.cache.loadDirectoryTreeSync(fullPath);
     }
 
-    loadDirectoryChildren(fullPath:string): Promise<(File | ShallowDirectory)[]> {
-        if (this.isTreeCached) return this.cache.loadDirectoryChildren(fullPath);
-        return this.cacheTree().then(() => {
-            this.isTreeCached = true;
-            return this.loadDirectoryChildren(fullPath);
-        });
+    async loadDirectoryChildren(fullPath:string): Promise<(File | ShallowDirectory)[]> {
+        if (this.isTreeCached) {
+            return this.cache.loadDirectoryChildrenSync(fullPath);
+        }
+
+        await this.cacheTree();
```

Comment:
out of scope of this PR, but I see now this is a de-opt introduced by myself as naive first implementation of `loadDirectoryChildren `.
I've opened an issue